### ISSUE TITLE
[SDL2] Unable to make EGL context current (call to eglMakeCurrent failed, reporting an error of EGL_BAD_ACCESS)

### ISSUE THREAD

**Issue by Raffaello** [libsdl-org/SDL#10023]
trying to create a openGL context and setting active on a different thread, the function `SDL_GL_MakeCurrent` returns `-1` with the error: `Unable to make EGL context current (call to eglMakeCurrent failed, reporting an error of EGL_BAD_ACCESS)`

This is happening on Linux with Wayland, using:
```
INFO: SDL Version: 2.30.3
INFO: Video Driver: wayland
DEBUG: Vendor:   Intel
DEBUG: Renderer: Mesa Intel(R) Graphics (ADL-N)
```

This is a minimum reproducible code (this code gives me the error on Fedora 40, but not on Windows 11):

```cpp
#include <SDL2/SDL.h>


SDL_GLContext glRenderContext;
SDL_GLContext glThreadContext;
SDL_Window*   mainWindow;
SDL_atomic_t  stuffIsReady;

int loadGlInBackground(void* ptr)
{
   // ********** ----> ERROR IS HAPPENING HERE  <----- ***********
    int res = SDL_GL_MakeCurrent(mainWindow, glThreadContext);
    if (res != 0)
    {
        SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION, "make current: %s (%d)", SDL_GetError(), res);
    }

    // Do some fancy loading gl stuff in the background...
    SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION, "Loading...");

    // It is very important in shared contexts to make sure the driver is done with all Objects before signaling other threads that they can use them!
    //GLsync fenceId = glFenceSync(GL_SYNC_GPU_COMMANDS_COMPLETE, 0);
    //GLenum result;
    //while (1)
    //{
        //SDL_Delay(1000);
        //glClientWaitSync(fenceId, GL_SYNC_FLUSH_COMMANDS_BIT, (GLuint64) 5000000000);    // 5 Second timeout
        //if (result != GL_TIMEOUT_EXPIRED)
            //break;                                                                       // we ignore timeouts and wait until all OpenGL commands are processed!
    //}

    SDL_AtomicIncRef(&stuffIsReady);

    // im not 100% sure what happens to a context that still is active when a thread gets closed.
    // But you can't call SDL_GL_DeleteContext on it!
    // So I just unbind it from this thread and call SDL_GL_DeleteContext in the main thread at closing.
    SDL_GL_MakeCurrent(mainWindow, NULL);
    return 0;
}

int main(int argc, char* argv[])
{
    if (SDL_Init(SDL_INIT_VIDEO) != 0)
    {
        SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION, "Unable to initialize SDL: %s", SDL_GetError());
        return -1;
    }
    else
    {
        SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION, "sdl init ok");
    }

    // SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 4);
    // SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 3);
    SDL_GL_SetAttribute(SDL_GL_SHARE_WITH_CURRENT_CONTEXT, 1);

    // SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);

    // SDL_GL_SetAttribute(SDL_GL_ACCELERATED_VISUAL, 1);
    // SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, 1);
    // SDL_GL_SetAttribute(SDL_GL_CONTEXT_FLAGS, 0);


    mainWindow = SDL_CreateWindow("SDL2 window", SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, 640, 480, SDL_WINDOW_SHOWN | SDL_WINDOW_OPENGL | SDL_WINDOW_RESIZABLE);

    if (mainWindow == NULL)
    {
        SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION, "Could not create window: %s", SDL_GetError());
        return -1;
    }

    // SDL_GL_SetAttribute(SDL_GL_SHARE_WITH_CURRENT_CONTEXT, 1);


    glRenderContext = SDL_GL_CreateContext(mainWindow);

    // if (!gladLoadGLLoader(SDL_GL_GetProcAddress))
    // {
    //     SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION, "can't load OpenGL extensions");
    //     return -1;
    // }

    glThreadContext = SDL_GL_CreateContext(mainWindow);
    SDL_GL_MakeCurrent(mainWindow, mainWindow);
    SDL_AtomicSet(&stuffIsReady, 0);
    SDL_Thread* thread = SDL_CreateThread(loadGlInBackground, "testThread", (void*) NULL);
    while (1)
    {
        // Event handling and all the fancy rendering...
        if (SDL_AtomicGet(&stuffIsReady) == 1)
        {
            // now the opengl stuff loaded by the thread can be used
            SDL_LogInfo(SDL_LOG_CATEGORY_APPLICATION, "Atomic One");
            break;
        }
    }

    int threadReturnValue;
    SDL_WaitThread(thread, &threadReturnValue);
    SDL_GL_DeleteContext(glThreadContext);
    SDL_GL_DeleteContext(glRenderContext);
    SDL_DestroyWindow(mainWindow);
    SDL_Quit();
    return 0;
}


```

**Comment by Raffaello** [comment]
with `export SDL_VIDEODRIVER=x11`  and then executing it looks like it is working (well not this code specifically but it doesn't fail anymore to set up the gl context)

**Comment by Raffaello** [comment]
with a better complex application i can confirm the problem looks likes and resides in the `wayland` video driver. 

with `x11` i am not experiencing any issue with OpenGL texture loading on a separated thread.

but with `wayland` there is the issue described above.

I am not sure is a problem of SDL2 wayland video driver neither at this point as i have looked into the SLD2 source code and it looks pretty straight forward that doesn't really allow to much of a bug into it...

Is there anything that might affect wayland and opengl contexts if anyone of you might known? thx

**Comment by madebr** [comment]
1. [`SDL_GL_MakeCurrent` expects a `SDL_GLContext`](https://wiki.libsdl.org/SDL2/SDL_GL_MakeCurrent), you're passing a `SDL_Window*`
2. There can only be one GL context active for a window at a time, so you need to do `SDL_GL_MakeCurrent(window, NULL)` in the main thread, before you can make it active in the subthread.

**Comment by Raffaello** [comment]
> 1. [`SDL_GL_MakeCurrent` expects a `SDL_GLContext`](https://wiki.libsdl.org/SDL2/SDL_GL_MakeCurrent), you're passing a `SDL_Window*`
> 2. There can only be one GL context active for a window at a time, so you need to do `SDL_GL_MakeCurrent(window, NULL)` in the main thread, before you can make it active in the subthread.

 thank you for the tips, i might look further, but
i am not sure to understand what you wrote:
1. at this line is using a `SDL_Window` and `SDL_GLContext`
```cpp
int res = SDL_GL_MakeCurrent(mainWindow, glThreadContext);
````
2. I am not sure this can be true otherwise there is no way to have a rendering thread on the main window and loading texture in the background with a different GL context (shared with the one in the main thread).

Finally the point of all of this is:

The code is working on Windows and Linux with X11 video driver. 
With Wayland doesn't.

----

p.s.
if you refere at this line
```
 SDL_GL_MakeCurrent(mainWindow, mainWindow);
 ```
apologies it might be a typo while writing the ticket or something.

-----

The fact remain that in Wayland doesn't work  `SDL_GL_MakeCurrent` on the thread reporting the error:
```
Unable to make EGL context current (call to eglMakeCurrent failed, reporting an error of EGL_BAD_ACCESS)
```

working instead with X11

**Comment by Kontrabant** [comment]
This is the same as #9072

You can't have two contexts bound to the same surface at one time. The EGL spec states:
> eglMakeCurrent returns EGL_FALSE on failure.
…
If ctx is current to some other thread, or if either draw or read are bound to contexts in another thread, an EGL_BAD_ACCESS error is generated.

Calling `SDL_GL_MakeCurrent(mainWindow, NULL);` to unbind the current context before starting the thread makes your example work.

**Comment by slouken** [comment]
Duplicate of https://github.com/libsdl-org/SDL/issues/9072

**Comment by Raffaello** [comment]
> This is the same as #9072
> 
> You can't have two contexts bound to the same surface at one time. The EGL spec states:
> 
> > eglMakeCurrent returns EGL_FALSE on failure.
> > …
> > If ctx is current to some other thread, or if either draw or read are bound to contexts in another thread, an EGL_BAD_ACCESS error is generated.
> 
> Calling `SDL_GL_MakeCurrent(mainWindow, NULL);` to unbind the current context before starting the thread makes your example work.

Yeah that it makes it work, but it won't render anything in the main thread. 


Is this a limitation of SDL2 that can't have a write-only gl context for background loading?
Would it be present in SDL3?

**Comment by Raffaello** [comment]
**jus for your information**:

I have found a workaround with Wayland video driver to render in the main window while uploading texture in background:

creating a Hidden borderless Window of size of 1 pixel with its own (shared) GL Context .

In this way, it is possible to upload sharable OpenGL objects (like textures) and render whatever on the main window (previous upload finished if it is the loading resource to be rendered).

The extra thing to do is to deal with `SDL_QUIT` event if needed as must be produced when the WINDOW EVENT CLOSE is produce for the main window.

Besides, that
It feels a bit bizarre to create a hidden window just to do loading resources while keep rendering in the  main window, but that's how to do it with Wayland.

**Comment by Raffaello** [comment]
Apologies if i come back to this again, but it is important:

looking at the code both main branch and release-2.28.x, the function `SDL_EGL_MakeCurrent` is used by the wayland implementation: `Wayland_GLES_MakeCurrent`

the  `eglMakeCurrent ` is defined as:
```cpp
eglMakeCurrent (EGLDisplay dpy, EGLSurface draw, EGLSurface read, EGLContext ctx);
```

As different from windows and x11, there is the 2 EGLSurfaces: `draw` and `read` and those 2 based on the EGL Specification when dealing with multiple contexts requires:

> If context is current to some other thread, or if either draw or read are bound to contexts in another thread, an EGL_BAD_ACCESS error is generated.

ref: https://github.com/libsdl-org/SDL/issues/9072#issuecomment-1947319731

The SDL Implemenation in `SDL_EGL_MakeCurrent` is using both read and draw with the same surface as noticed here:
```cpp
if (!_this->egl_data->eglMakeCurrent(_this->egl_data->egl_display,
                                             egl_surface, egl_surface, egl_context)) {
            return SDL_EGL_SetError("Unable to make EGL context current", "eglMakeCurrent");
        }
```

And i think in the specific case of this ticket  the background loading thread instead should make like: `egLMakeCurrent(_this->egl_data->egl_display, EGL_NO_SURFACE, egl_surface, egl_context)` so basically without setting the `draw` surface.

This is not possible by the current public SDL APIs (i think neither for the SDL3 implmentation).

would it be possible to think out something to give this extra flexibility to the users? 

the first idea that came to my mind now it would be having a `SDL_MakeCurrentEx` where both read and draw context can be specified, in the case the underneath video driver is not an EGL related,  it will only use the draw and ignoring the read and call the normal `SDL_MakeCurrent` or something like that
